text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+  receiver_inputs = []
+  receiver_outputs = []
+  // receive-only wallets could have receive_protected_prob=0.0 to always have 1-2 receive transactions
+  is_protected = random() <= config.RECEIVE_PROTECTED_PROB
+  if is_protected {
+    receiver_inputs, receiver_outputs = receive_protected()
+    receiver_outputs += [generate_protected_output(value)]
+  } else {
+    receiver_outputs += [generate_unprotected_output(value)]
+  }
+  return receiver_inputs, receiver_outputs
+}
+
+// A 'send' transaction needs to always be protected from replays so we MUST
+// either add a protected output as an input or make it a bootstrap transaction
+// if an anchor does not exist yet
+fn send(value: int64) -> ([]Output, []Output) {
+  // first pick the inputs that hold enough value to perform a tx
+  inputs = pick_random_inputs()
+  outputs = []
+
+  // if we have no anchor, contribute anchor + PS outputs for replay protection
+  if anchor_not_exists() {
+    outputs = bootstrap_outputs()
+  // otherwise, add a protected output as input to protect the transaction
+  } else {
+    // contribute a protected output as an input if it's not already contributed
+    if count_protected_output(inputs) == 0 {
+        inputs += [random_protected_output()]
+    }
+  }
+  change_value = sum(input.v for input in inputs) - value  // fee is missing from here
+  // we know our change output will be protected either by bootstrap outputs or by
+  // a protected input so we label it as a protected output
+  change_output = generate_protected_output(change_value)
+  outputs += [change_output]
+  return inputs, outputs
+}
+
+// TODO: needs handling of running out of protected outputs and keeping a minimal of `N`
+// protected outputs as defined in the configuration file
+```
+
+TODO: Check this is safe.
+
+#### Protection with wallet history
+
+A wallet can remember which outputs were spent. This way, if a spent output reappears in the wallet, a user is given a choice to either accept it or refresh it through a self-spend transaction.
+
+The downside of this approach is that a replay attacks are still possible in which case it would mean that a user would need to make a manual choice what to do with outputs that are unknown to the history of the wallet. Unknown outputs are also those created from the same seed on a different device and from child wallets since they each have their own history. The cross device scenario could be mitigated by export and import of wallet history and the child wallet outputs can be labeled and hence assumed as safe. It's up to the child wallet to protect itself from the attacks.
+
+### Receive-only wallets
+ 
+Any kind of automated receiving should default to 1-2 transactions and thus creating _unprotected_ outputs to avoid utxo spoofing attack which would reveal our inputs. Always performing 1-2 receive transactions can be achieved by setting the configuration `RECEIVE_PROTECTED_PROB = 0.0` which is explained in the next section.

```

Comment:
It's true that a receiver is more at risk of utxo spoofing; a receiver has an inherent incentize to see a tx confirm. The receiver cannot be spoofed if they are the one to finalize a tx.
I remember thinking a while back that only sender finalization makes sense, but now I cannot the recall why that should be.
Anyway, with the possibility of receiver finalization, they should absolutely do payjoins to receive. This puts the utxo spoofing risk on the sender, but the sender has already determined that the receiver is worthy of being paid, so they probably perceive a much smaller utxo spoofing risk.

Reply:
Yeah, when I wrote this, I had in mind the current transaction building scheme. I agree about the payjoins in the receiver finalizing the tx scenario :+1: One thing i'm still a bit concerned about is the automated payjoins (I described the reasoning here https://github.com/mimblewimble/grin-rfcs/pull/59#discussion_r468042139)